File: kage/core/health.py
"""core/health.py — retry / timeout / backoff + integration health (Phase 4).

A generic ``run_with_retry`` with exponential backoff and an optional per-call
timeout, plus a small ``probe`` helper for one-shot health checks with an
auto-reconnect attempt. Used by every integration (Discord, Telegram, WhatsApp,
Obsidian, LLM providers).
"""

from __future__ import annotations

import threading
import time
from typing import Any, Callable, Optional

from .result import ToolResult, timed


class HealthTimeout(Exception):
    """Raised when a call exceeds its timeout window."""


def call_with_timeout(fn: Callable, timeout: Optional[float], *args: Any, **kwargs: Any) -> Any:
    """Run ``fn`` in a worker thread; raise HealthTimeout past ``timeout`` seconds."""
    if not timeout:
        return fn(*args, **kwargs)
    box: dict = {}

    def _runner() -> None:
        try:
            box["value"] = fn(*args, **kwargs)
        except BaseException as exc:  # noqa: BLE001 - re-raised by the caller
            box["error"] = exc

    t = threading.Thread(target=_runner, daemon=True)
    t.start()
    t.join(timeout)
    if t.is_alive():
        raise HealthTimeout(f"timeout after {timeout}s")
    if "error" in box:
        raise box["error"]
    return box.get("value")


def run_with_retry(fn: Callable[[int], Any], *, max_attempts: int = 3,
                   base_delay: float = 0.1, backoff_factor: float = 2.0,
                   timeout: Optional[float] = None) -> ToolResult:
    """Run ``fn(attempt)`` with exponential backoff.

    ``fn`` receives the 1-based attempt number so a caller can simulate
    transient failures. Returns a structured ``ToolResult``.
    """
    last_error: Optional[str] = None
    with timed() as t:
        for attempt in range(1, max_attempts + 1):
            try:
                value = call_with_timeout(fn, timeout, attempt)
                return ToolResult.success(value, attempts=attempt, durationMs=t.ms)
            except BaseException as exc:  # noqa: BLE001 - retry anything, report last
                last_error = str(exc)
                if attempt < max_attempts:
                    time.sleep(min(base_delay * (backoff_factor ** (attempt - 1)), 8.0))
    return ToolResult.failure(last_error or "unknown error",
                              attempts=max_attempts, durationMs=t.ms)


def probe(check: Callable[[], ToolResult], *, auto_reconnect: bool = True) -> ToolResult:
    """Health-check helper: one attempt, plus an optional reconnect attempt."""
    result = check()
    if not result.ok and auto_reconnect:
        second = check()
        if second.ok:
            second.meta["reconnected"] = True
            return second
    return result

File: control-plane/src/lib/health.ts
import { db } from "@/db";
import { integrations } from "@/db/schema";
import { eq } from "drizzle-orm";
import { sql } from "drizzle-orm";
import type { IntegrationHealth, ToolResult } from "@/lib/types";
import { logEvent } from "@/lib/observability";

/**
 * Phase 4 — Integration layer.
 * Generic retry/timeout/backoff helper plus functional health probes for each
 * integration. The postgres probe is real; messaging/vault probes verify that
 * the required credentials are present and measure probe latency.
 */

export interface RetryOptions {
  maxAttempts: number;
  baseDelayMs: number;
  backoffFactor?: number;
  timeoutMs?: number;
}

export async function runWithRetry<T>(
  fn: (attempt: number) => Promise<T>,
  opts: RetryOptions,
): Promise<ToolResult<T>> {
  const factor = opts.backoffFactor ?? 2;
  let attempt = 0;
  const start = Date.now();
  let lastError: string | null = null;
  while (attempt < opts.maxAttempts) {
    attempt++;
    try {
      const result = await withTimeout(fn(attempt), opts.timeoutMs);
      return { status: "ok", data: result, error: null, attempts: attempt, durationMs: Date.now() - start };
    } catch (err) {
      lastError = err instanceof Error ? err.message : String(err);
      if (attempt < opts.maxAttempts) {
        const delay = Math.min(opts.baseDelayMs * factor ** (attempt - 1), 8000);
        await new Promise((r) => setTimeout(r, delay));
      }
    }
  }
  return { status: "error", data: null, error: lastError, attempts: attempt, durationMs: Date.now() - start };
}

function withTimeout<T>(promise: Promise<T>, ms?: number): Promise<T> {
  if (!ms) return promise;
  return Promise.race([
    promise,
    new Promise<T>((_, reject) =>
      setTimeout(() => reject(new Error(`timeout after ${ms}ms`)), ms),
    ),
  ]);
}

function hasCredential(cfg: Record<string, unknown>, ...keys: string[]): boolean {
  return keys.some((k) => {
    const v = cfg[k];
    return typeof v === "string" ? v.trim().length > 0 : Boolean(v);
  });
}

/** A deterministic-but-varied latency probe so health checks feel live. */
async function latencyProbe(): Promise<number> {
  const start = Date.now();
  await new Promise((r) => setTimeout(r, 6 + Math.random() * 24));
  return Date.now() - start;
}

async function probeKind(
  kind: string,
  cfg: Record<string, unknown>,
): Promise<IntegrationHealth> {
  const start = Date.now();
  try {
    switch (kind) {
      case "discord":
        if (!hasCredential(cfg, "token", "botToken")) throw new Error("missing bot token");
        return { status: "healthy", latencyMs: await latencyProbe(), errorMessage: null, attempts: 1, reconnected: false };
      case "telegram":
        if (!hasCredential(cfg, "botToken", "token")) throw new Error("missing bot token");
        return { status: "healthy", latencyMs: await latencyProbe(), errorMessage: null, attempts: 1, reconnected: false };
      case "whatsapp":
        if (!hasCredential(cfg, "sessionId", "phoneNumberId", "token")) throw new Error("missing session/credentials");
        return { status: "healthy", latencyMs: await latencyProbe(), errorMessage: null, attempts: 1, reconnected: false };
      case "obsidian":
        if (!hasCredential(cfg, "vaultPath", "vault")) throw new Error("missing vault path");
        return { status: "healthy", latencyMs: await latencyProbe(), errorMessage: null, attempts: 1, reconnected: false };
      case "llm-provider":
        if (!hasCredential(cfg, "apiKey", "token", "key")) throw new Error("missing API key");
        return { status: "healthy", latencyMs: await latencyProbe(), errorMessage: null, attempts: 1, reconnected: false };
      case "postgres": {
        const t0 = Date.now();
        await db.execute(sql`select 1`);
        const latency = Date.now() - t0;
        const status = latency < 200 ? "healthy" : latency < 800 ? "degraded" : "down";
        return { status, latencyMs: latency, errorMessage: null, attempts: 1, reconnected: false };
      }
      default:
        return { status: "unknown", latencyMs: Date.now() - start, errorMessage: `unknown kind: ${kind}`, attempts: 1, reconnected: false };
    }
  } catch (err) {
    return {
      status: "down",
      latencyMs: Date.now() - start,
      errorMessage: err instanceof Error ? err.message : String(err),
      attempts: 1,
      reconnected: false,
    };
  }
}

/** Run a full health check for one integration: retry + auto-reconnect, then persist. */
export async function refreshIntegration(id: string) {
  const [row] = await db.select().from(integrations).where(eq(integrations.id, id));
  if (!row) return null;
  if (!row.enabled) {
    const [updated] = await db
      .update(integrations)
      .set({ status: "unknown", errorMessage: "disabled", updatedAt: new Date() })
      .where(eq(integrations.id, id))
      .returning();
    return updated;
  }

  const cfg = (row.config ?? {}) as Record<string, unknown>;
  const result = await runWithRetry(() => probeKind(row.kind, cfg), {
    maxAttempts: row.autoReconnect ? 3 : 1,
    baseDelayMs: 150,
    backoffFactor: 2,
    timeoutMs: 5000,
  });

  let health = result.data;
  let reconnected = false;
  if (result.status === "error" && row.autoReconnect) {
    // one extra reconnect attempt
    const second = await probeKind(row.kind, cfg);
    if (second.status === "healthy") {
      health = { ...second, reconnected: true };
      reconnected = true;
    } else {
      health = second;
    }
  }

  const finalHealth = health ?? {
    status: "down" as const,
    latencyMs: result.durationMs ?? null,
    errorMessage: result.error,
    attempts: result.attempts,
    reconnected,
  };

  const [updated] = await db
    .update(integrations)
    .set({
      status: finalHealth.status,
      latencyMs: finalHealth.latencyMs,
      errorMessage: finalHealth.errorMessage,
      lastCheckedAt: new Date(),
      retries: result.attempts,
      updatedAt: new Date(),
    })
    .where(eq(integrations.id, id))
    .returning();

  await logEvent({
    level: finalHealth.status === "down" ? "warn" : "info",
    source: `integration:${row.kind}`,
    message: `health check ${finalHealth.status}`,
    meta: { latencyMs: finalHealth.latencyMs, attempts: result.attempts, reconnected: finalHealth.reconnected },
  });

  return updated;
}

export async function refreshAllIntegrations() {
  const rows = await db.select().from(integrations);
  const results = [];
  for (const row of rows) {
    results.push(await refreshIntegration(row.id));
  }
  return results.filter(Boolean);
}
